mammoth/inputters/dataloader.py

```diff
@@ -86,40 +103,75 @@ def __init__(self, examples_stream, look_ahead_size, n_buckets, batch_size, buck
     def _init(self):
         logger.info('LookAheadBucketing: initialization start')
         for example in itertools.islice(self.examples_stream, self.look_ahead_size):
-            bucket_idx = self.bucket_fn(example)
-            self._buckets[bucket_idx].append(example)
-            self._lens[bucket_idx] += 1
+            s_bucket, t_bucket = self.bucket_fn(example)
+            self._buckets[s_bucket][t_bucket].append(example)
+            self._lens[s_bucket][t_bucket] += 1
         logger.info('LookAheadBucketing: initialization done')
 
     def maybe_replenish(self) -> bool:
         """look up one more example to add to this reservoir."""
         try:
             example = next(self.examples_stream)
-            bucket_idx = self.bucket_fn(example)
-            creates_new_bucket = self._lens[bucket_idx] == 0
-            self._buckets[bucket_idx].append(example)
-            self._lens[bucket_idx] += 1
+            s_bucket, t_bucket = self.bucket_fn(example)
+            creates_new_bucket = self._lens[s_bucket][t_bucket] == 0
+            self._buckets[s_bucket][t_bucket].append(example)
+            self._lens[s_bucket][t_bucket] += 1
             return creates_new_bucket
         except StopIteration:
             return None
 
-    def bucket_is_empty(self, bucket_idx) -> bool:
-        return self._lens[bucket_idx] == 0
+    def bucket_is_empty(self, s_idx, t_idx) -> bool:
+        return self._lens[s_idx][t_idx] == 0
 
     def _choose_and_prepare_bucket(self, bucket_idx=None):
         """pick a bucket (at random unless specified) and prepare examples for iteration"""
         if bucket_idx is None:
-            bucket_idx = random.choices(range(len(self._buckets)), weights=self._lens, k=1)[0]
+            buckets = [(s, t) for s in range(self.n_buckets) for t in range(self.n_buckets)]
+            weights = [self._lens[s][t] for s in range(self.n_buckets) for t in range(self.n_buckets)]
+            s_bucket, t_bucket = random.choices(buckets, weights=weights, k=1)[0]
         # if bucket_idx >= len(self._buckets):
         #     import pdb; pdb.set_trace()
         # if len(self._prefetched[self._buckets[bucket_idx]]) == 0:
         #     import pdb; pdb.set_trace()
-        random.shuffle(self._buckets[bucket_idx])
-        return bucket_idx
+        random.shuffle(self._buckets[s_bucket][t_bucket])
+        return s_bucket, t_bucket
 
     def is_empty(self):
         return all(size == 0 for size in self._lens)
```

Comment:
This also needs to be updated to account for `self._lens` being twodimensional.
Currently this will always be false.